### ISSUE TITLE
Large import from OneTab hung and created duplicate entries - Multi-Account Containers issue?

### ISSUE THREAD

**Issue by codebykat** [josh-berry/tab-stash#91]
Hi! I'm trying to switch from OneTab, where I have quite a few tabs stored. The import was proceeding fine (if, slowly) but it seems to have stalled out. No errors in the web console (but I didn't have it open before the import stalled out).

I can also no longer cancel the import - I had to reload the Tab Stash tab.

The import opened a LOT of tabs which I guess I just need to stash manually now, and also created quite a few duplicate entries -- both under the same group and different groups. I did double-check that these URLs aren't duplicated in the export from OneTab, and they aren't something silly like pages that require authentication and redirect to a login page.

In typing this out, I realized the issue might be related to the Multi-Account Containers extension. It seems that all the lingering tabs left open are sites I have set to "always open" in a different container than the one that Tab Stash is open in.

I'm also not sure if the entries are true duplicates or some other kind of bug. I just opened a new tab and it seems the sidebar is reflecting many copies of whatever tab I am currently looking at.

### System Details
- **OS Version:** OSX Catalina 10.15.6
- **Browser Version:** Firefox 80.0.1
- **Tab Stash Version:** 2.6

Among others, I have the following plugins installed: OneTab, Multi-Account Containers

### Describe the Bug

Import is no longer progressing nor reporting any errors; many tabs are open and remaining open; duplicates have been created.

<img width="690" alt="Screen Shot 2020-09-23 at 2 01 39 AM" src="https://user-images.githubusercontent.com/52152/93991389-15e12680-fd41-11ea-9a58-05e4a2074f91.png">

<img width="194" alt="Screen Shot 2020-09-23 at 2 14 09 AM" src="https://user-images.githubusercontent.com/52152/93992494-7a50b580-fd42-11ea-9735-bcdfe1060506.png">

<img width="204" alt="Screen Shot 2020-09-23 at 2 16 04 AM" src="https://user-images.githubusercontent.com/52152/93992732-c00d7e00-fd42-11ea-8f09-3e9386149ab3.png">

### Expected Behavior
Expected import to complete, or to properly handle errors.

### Steps to Reproduce
1. Have mult-account containers set up with some sites set to "always open" in a given container.
2. In OneTab, Export and copy the provided text, including URLs to the above sites.
3. Paste this list of URLs into Tab Stash and leave the settings at default (break on paragraphs and headers, if I recall correctly).

### Impact
I really want to try Tab Stash but I have a LOT of stored data in OneTab so I feel kind of locked in. The reason I installed Tab Stash to try it was because I saw there's a OneTab importer.

Secondarily, I installed a tab management extension because I wanted to get my tab situation under control, and now I have 139 open tabs 😱  some of them months old 🙈

<hr>
<em>Readers: If you are also experiencing this bug, please vote for it by giving the ORIGINAL POST a thumbs-up using the :smiley: button.  You are welcome to leave comments and discuss the issue if you have more details to add, but "Me too!" comments are not counted by the voting system.</em>

**Comment by josh-berry** [comment]
Hi, sorry for the trouble, and thanks for taking the time to write up a detailed bug report!

I installed Multi-Account Containers just now and tested your scenario and I see the same issue. I did a little more digging and agree it's related to the "always open" option. Here's what I think might be happening:

During the import process, Tab Stash creates some temporary tabs which it uses to fetch the page title and icon for each imported URL. Under normal circumstances, it will open a limited number of tabs (4) and reuse them so as not to overload your browser.

I suspect Multi-Account Containers is noticing that one of these temporary tabs (which is in the default container) was navigating to a site that is set to "always open" in another container, and it is closing that tab and opening a new one in the right container. This in turn is confusing Tab Stash, which doesn't expect one of its temporary tabs to be closed--so Tab Stash is listening for events on the closed tab which will never happen, and the import process freezes.

--------------

Unfortunately I'm not sure of a good way to work around this right now, other than to disable Multi-Account Containers just for the import (which I know is not great since it compromises on the privacy/security benefits of using containers in the first place).

Once the import is done, however, I don't expect you'll have any problems actually *using* Tab Stash--the importer (and the "Fetch Missing Icons" command which uses the same code) is I think the only place where Tab Stash tries to use the same tab for a long-running operation. (But please feel free to file more bugs if you notice any other problems!)

So if you're comfortable with temporarily disabling containers during the import, that would be my suggestion for now, and I'll try to look into this further to see if I can handle it more gracefully when a "temporary" tab disappears (due to Multi-Account Containers or any other reason).

Thanks again for the bug report and sorry for the inconvenience!

**Comment by codebykat** [comment]
Thanks for the reply! Yes, I've been using it today and it's really delightful. The sidebar especially, with the search field, is a huge UI improvement over OneTab. I haven't seen any other conflicts with the containers.

I basically just re-stashed all the open container tabs and then sorted through them. In the end it doesn't appear anything was lost, though I had to copy and paste a few URLs (as the item in Tab Stash had the name of the missing URL, but the URL of Tab Stash).

I don't suppose you can detect if another extension is installed. Maybe the text on the import box could add an instruction to disable the containers extension before importing? Unfortunately Multi-Account Containers do seem to have trouble playing nice with quite a few other extensions.

**Comment by josh-berry** [comment]
> Thanks for the reply! Yes, I've been using it today and it's really delightful. The sidebar especially, with the search field, is a huge UI improvement over OneTab. I haven't seen any other conflicts with the containers.

Thank you, that's great to hear!

> I don't suppose you can detect if another extension is installed. Maybe the text on the import box could add an instruction to disable the containers extension before importing? Unfortunately Multi-Account Containers do seem to have trouble playing nice with quite a few other extensions.

You're right that Firefox doesn't let extensions know if other extensions are installed (at least, not without asking for additional permissions). If I can, I'm hoping I can find a way to detect and gracefully handle the behavior (temporary tab closing and new container-tab opening) so import will just work even with Multi-Account Containers. But if I can't do that, then I at least want to have the import fail gracefully and explain to the user what happened.

**Comment by zouglou** [comment]
Hello,

I've kinda the same problem, but I don't use Multi-Account Containers, so no "always opened" tabs, but I use Facebook Container (and Conex, but no such settings I believe) .

I started an import of ~200 tabs from OneTab, seeing it would take a bit of time, went to the kitchen, and when I came back, the import was mostly done (~90%) but seemed to be stopped.
I first thought it was caused by Auto Tab Discard discarding tabs after 360s of inactivity: when I came back, there was only one tab processing and it was discarded, but I didn't know if it got discarded before or after import failure.
Clicking on the Cancel button had no effect at this moment.

So I tried back, with Auto Tab Discard deactivated, and I got kinda the same result: same tab waiting here (the only facebook tab), and a second still loading (just reactivated JS (using NoScript) on it, and it immediately got stashed).
Clicking on Cancel has still no effect.

I'm going to retry with Facebook Container deactivated and I tell you what happen.

**Comment by josh-berry** [comment]
Yeah, it's very likely Facebook Container is doing the same thing that Multi-Account Containers does (where it closes and re-opens the tab in the correct container).

I'm hoping the workaround I have in mind will work for both of these extensions (and any other extension that re-opens tabs to move them between containers). But thanks for letting me know about Facebook Container too, I'll update the wiki page.

**Comment by josh-berry** [comment]
Alright, I've put in a fix/workaround for importing which will notice when a tab has been closed by Multi-Account Containers (or any other such extension), and try to continue the import using the new tab. Once the next release (2.7) is out, please give it a try and let me know if you run into any issues.

Thanks again for the report and sorry for the inconvenience!

**Comment by codebykat** [comment]
So nice to have an extension actively maintained and with such a care for the little details! Love it! I also really enjoyed the bit about the cats 😁 

<img width="451" alt="Screen Shot 2020-11-08 at 11 49 03 AM" src="https://user-images.githubusercontent.com/52152/98482803-9e643b00-21b8-11eb-850b-7c1cbc16647b.png">

**Comment by josh-berry** [comment]
Nice, you found the Easter egg! 😹  And thank you so much for the kind words, it means a lot to hear. 😄 

Of course, right after I released 2.7, I found another issue with import (and so did someone else, see #96). I'm going to push out a fix ASAP but wanted to mention it here as well just in case you happen to try out import again and run into it.
